### PR TITLE
adding collection of maps support

### PR DESCRIPTION
**Pull request by KennethNickles**
Not a pretty change (ie all the unbounded generics) but I needed to add support for `Collection<Map<String, Object>>` Open to suggestions, but this seems to be passing my tests. Prior to my changes autoparse parses the below as `Collection<UnknownObject>`

`Could not convert value in array at "myCollectionOfMapsOfStrings" to java.util.Map from org.json.JSONObject.`

```
  "myCollectionOfMapsOfStrings": [
    {
      "key1": "object 1",
      "key2": "object 2"
    }
  ],
  "myCollectionOfMapsOfObjects": [
    {
      "key1": {
        "object": "testObject"
      },
      "key2": {
        "object": "testObject"
      }
    }
  ],
  ```

### CODE REVIEW

File: core/src/main/java/com/workday/autoparse/json/parser/JsonParserUtils.java [hunk 1]
```diff
@@ -781,13 +781,29 @@ public static <T> void parseJsonArray(JsonReader reader,
...
             while (reader.hasNext()) {
-                Collection nextCollection = nextCollectionInitializer.newInstance();
-                parseJsonArray(reader,
-                               nextCollection,
-                               itemParser,
-                               itemType,
-                               innerCollectionClasses.subList(1, innerCollectionClasses.size()),
-                               key);
+                Object nextCollection = nextCollectionInitializer.newInstance();
+                if (nextCollection instanceof Collection) {
```
Comment (plattysoft) [inline core/src/main/java/com/workday/autoparse/json/parser/JsonParserUtils.java:785]: I think we need to rename all the functions and variables that refer to `collections` this code feels really strange now:
```java
if (nextCollection instanceof Collection) {
   ...
} else if ...
```
If it is called a collection, it should be a collection (same goes for the `CollectionInitializer` and `getCollectionInitializerForClass`)

File: demo/src/test/java/com/workday/autoparse/json/demo/JsonParserTest.java [hunk 2]
```diff
@@ -189,6 +191,18 @@ private void testParse(String fileName) throws Exception {
         assertEquals("testObject.collectionOfCollection[1]", CollectionUtils.newArrayList(3, 4),
                      collectionOfCollections.get(1));
 
+        List<Map<String, String>> collectionOfMapsOfStrings = testObject.myCollectionOfMapsOfStrings;
```
Comment (plattysoft) [inline demo/src/test/java/com/workday/autoparse/json/demo/JsonParserTest.java:194]: This test is great!

File: core/src/main/java/com/workday/autoparse/json/initializers/CollectionInitializerFactory.java [hunk 3]
```diff
@@ -39,7 +42,13 @@ public static CollectionInitializer getCollectionInitializerForClass(
             return HashSetInitializer.INSTANCE;
         } else if (LinkedHashSet.class.equals(clazz)) {
             return LinkedHashSetInitializer.INSTANCE;
-        } else {
+        } else if (Map.class.equals(clazz)) {
+            return HashMapInitializer.INSTANCE;
+        } else if (HashMap.class.equals(clazz)) {
+            return HashMapInitializer.INSTANCE;
+        } else if (LinkedHashMap.class.equals(clazz)) {
+            return LinkedHashMapInitializer.INSTANCE;
```
Comment (plattysoft) [inline core/src/main/java/com/workday/autoparse/json/initializers/CollectionInitializerFactory.java:50]: Maybe we could just have `Map` here. Allowing the specifics of LinkedHashMap and HashMap seems unnecessary,  IMHO the data class that receives this only should use the interface `Map`

**Review by plattysoft — APPROVED** [review]
Not pretty but works. I left a couple of suggestions to improve readability.

File: core/src/main/java/com/workday/autoparse/json/initializers/CollectionInitializerFactory.java (hunk 3 above)
Comment (KennethNickles) [reply core/src/main/java/com/workday/autoparse/json/initializers/CollectionInitializerFactory.java:50]: These were the supported map types I gleaned from other classes. So I thought it prudent to allow the same types in collections.

File: core/src/main/java/com/workday/autoparse/json/parser/JsonParserUtils.java (hunk 1 above)
Comment (KennethNickles) [reply core/src/main/java/com/workday/autoparse/json/parser/JsonParserUtils.java:785]: Do you have a suggestion for something that encompasses Map and Collection?

Comment (plattysoft) [reply core/src/main/java/com/workday/autoparse/json/parser/JsonParserUtils.java:785]: How does group sounds?

File: core/src/main/java/com/workday/autoparse/json/initializers/CollectionInitializerFactory.java (hunk 3 above)
Comment (plattysoft) [reply core/src/main/java/com/workday/autoparse/json/initializers/CollectionInitializerFactory.java:50]: fair

File: core/src/main/java/com/workday/autoparse/json/parser/JsonParserUtils.java (hunk 1 above)
Comment (KennethNickles) [reply core/src/main/java/com/workday/autoparse/json/parser/JsonParserUtils.java:785]: I think I'll do this in another PR. It has potential to break the code gen.

**Pull request merged**